File: backend/app/schemas/analytics.py
from pydantic import BaseModel


class VocabularyGrowthPoint(BaseModel):
    date: str
    count: int
    cumulative: int


class ReviewActivityPoint(BaseModel):
    date: str
    count: int


class ReviewQualityPoint(BaseModel):
    date: str
    again: int
    hard: int
    medium: int
    easy: int


class MasteryTrendPoint(BaseModel):
    date: str
    level_0: int
    level_1: int
    level_2: int
    level_3: int
    level_4: int
    level_5: int


class WritingInsightPoint(BaseModel):
    date: str
    grammar: int
    wording: int


class AnalyticsSummary(BaseModel):
    vocabulary_growth: list[VocabularyGrowthPoint]
    review_activity: list[ReviewActivityPoint]
    review_quality: list[ReviewQualityPoint]
    mastery_trend: list[MasteryTrendPoint]
    writing_insights: list[WritingInsightPoint]


class DashboardWidgetsUpdate(BaseModel):
    widgets: list[str]

File: frontend/src/types/analytics.ts
export type AnalyticsRange = "7d" | "30d" | "90d" | "all";

export interface VocabularyGrowthPoint {
  date: string;
  count: number;
  cumulative: number;
}

export interface ReviewActivityPoint {
  date: string;
  count: number;
}

export interface ReviewQualityPoint {
  date: string;
  again: number;
  hard: number;
  medium: number;
  easy: number;
}

export interface MasteryTrendPoint {
  date: string;
  level_0: number;
  level_1: number;
  level_2: number;
  level_3: number;
  level_4: number;
  level_5: number;
}

export interface WritingInsightPoint {
  date: string;
  grammar: number;
  wording: number;
}

export interface AnalyticsSummary {
  vocabulary_growth: VocabularyGrowthPoint[];
  review_activity: ReviewActivityPoint[];
  review_quality: ReviewQualityPoint[];
  mastery_trend: MasteryTrendPoint[];
  writing_insights: WritingInsightPoint[];
}

export type WidgetId =
  | "vocabulary_growth"
  | "review_activity"
  | "review_quality"
  | "mastery_trend"
  | "writing_insights";

export const WIDGET_LABELS: Record<WidgetId, string> = {
  vocabulary_growth: "Vocabulary Growth",
  review_activity: "Review Activity",
  review_quality: "Review Quality",
  mastery_trend: "Mastery Trend",
  writing_insights: "Writing Insights",
};

export const ALL_WIDGET_IDS: WidgetId[] = [
  "vocabulary_growth",
  "review_activity",
  "review_quality",
  "mastery_trend",
  "writing_insights",
];
